File: linked_list.c
#include <stdio.h>
#include <stdlib.h>


typedef struct Node {
    void* data;
    struct Node* next;
} Node;


typedef struct {
    Node* head;
    size_t size;
} LinkedList;    


Node* create_node(void* data);
void init_list(LinkedList* list);
int size(LinkedList *list);
int empty(LinkedList* list); 
void push_front(LinkedList* list, void* data);
void push_back(LinkedList* list, void* data); 
void* pop_front(LinkedList* list); 
void* pop_back(LinkedList* list); 
void* front(LinkedList* list); 
void* back(LinkedList* list); 
void insert(LinkedList* list, size_t index, void* data); 
void* erase(LinkedList* list, size_t index);
void* value_n_from_end(LinkedList* list, size_t n);
void reverse(LinkedList* list);
void* remove_value(LinkedList* list, void* value);
void free_list(LinkedList* list); 



Node* create_node(void* data) {
    Node* node = malloc(sizeof(*node));
    if (!node) {
        printf("Memory allocation failed\n");
        exit(1);
    }

    node->data = data;
    node->next = NULL;
    return node;
}


void init_list(LinkedList* list) {
    list->head = NULL;
    list->size = 0;
}


int size(LinkedList *list) {
    return list->size;
}


int empty(LinkedList* list) {
    return list->size == 0;
}


void push_front(LinkedList* list, void* data) {
    Node* node = create_node(data); 
    node->next = list->head;
    list->head = node;
    list->size++;
}


void push_back(LinkedList* list, void* data) {
    Node* node = create_node(data);
    if (empty(list)) {
        list->head = node;
    } else {
        Node* current = list->head;
        while (current->next) {
            current = current->next;
        }
        current->next = node;
    }

    list->size++;
}


void* pop_front(LinkedList* list) {
    if (!list->head) {
        return NULL;
    }

    Node* temp = list->head;
    void* data = temp->data;

    list->head = temp->next;
    free(temp);
    list->size--;
    return data;
}


void* pop_back(LinkedList* list) {
    if (!list || !list->head) {
        return NULL;
    }

    Node* current = list->head; 

    if (current->next == NULL) {
        void* data = current->data;
        free(current);
        list->head = NULL;
        list->size--;
        return data;
    }

    while (current->next->next != NULL) {
        current = current->next;
    }

    Node* last = current->next;
    void* data = last->data;

    current->next = NULL;
    free(last);
    list->size--;

    return data;

}


void* front(LinkedList* list) {
    if (!list || !list->head) {
        return NULL;
    }

    return list->head->data;
}


void* back(LinkedList* list) {
    if (!list || !list->head){
        return NULL;
    }

    Node* current = list->head;
    while (current->next) {
        current = current->next;
    }
    return current->data;

}


void insert(LinkedList* list, size_t index, void* data) {
    if (!index || index > list->size) {
        printf("Index out of range");
        exit(1);
    }
    if (index == 0) {
       push_front(list, data);
       return;
    }
    Node* node = create_node(data);
    Node* current = list->head;

    for (size_t i = 0; i < index - 1; i++) {
        current = current->next;
    }

    node->next = current->next;
    current->next = node;
    list->size++;
}

     
void* erase(LinkedList* list, size_t index) {
    if (!list || index >= list->size) {
        printf("Index out of range");
        exit(1);
    }
    if (index == 0) {
        return pop_front(list);
    } 
    Node* current = list->head;

    for (size_t i = 0; i < index - 1; i++) {
        current = current->next;
    }

    Node* target = current->next;
    void* data = target->data;

    current->next = target->next;
    free(target);
    list->size--;

    return data;
}

        
void* value_n_from_end(LinkedList* list, size_t n) {
    if (empty(list)) {
        return NULL;
    }
    if (n >= list->size) {
        printf("n is too large\n");
        exit(1);
    }

    Node* fast = list->head;
    Node* slow = list->head;

    for (size_t i = 0; i < n; i++) {
        fast = fast->next;
    }

    while (fast->next) {
        slow = slow->next;
        fast = fast->next;
    }

    return slow->data;
}
        

void reverse(LinkedList* list) {
    if (empty(list)) {
        return;
    }

    Node* prev = NULL;
    Node* current = list->head;

    while (current) {
        Node* next_node = current->next;
        current->next = prev;
        prev = current;
        current = next_node;
    }

    list->head = prev;
}


void* remove_value(LinkedList* list, void* value) {
    if (!list || !list->head) {
        return NULL;
    }

    Node* current = list->head;
    Node* prev = NULL;

    while (current) {
        if (current->data == value) {
            void* data = current->data;

            if (prev == NULL) {
                list->head = current->next;
            } else {
                prev->next = current->next;
            }
            
            free(current);
            list->size--;
            return data;
        }

        prev = current;
        current = current->next;
    }

    return NULL; 
}


void free_list(LinkedList* list) {
    Node* current = list->head;

    while (current) {
        Node* temp = current;
        current = current->next;
        free(temp);
    }

    list->head = NULL;
    list->size = 0;
}

int main() {
    LinkedList list;
    init_list(&list);

    printf("=== Testing push_back ===\n");
    for (int i = 1; i <= 5; i++) {
        int* value = malloc(sizeof(int));
        *value = i;
        push_back(&list, value);
    }

    printf("Size after push_back: %lu\n", (unsigned long)list.size);

    printf("\n=== Testing insert at index 2 (value 99) ===\n");
    int* special = malloc(sizeof(int));
    *special = 99;
    insert(&list, 2, special);

    printf("Size after insert: %lu\n", (unsigned long)list.size);

    printf("\n=== Printing list ===\n");
    Node* current = list.head;
    while (current) {
        printf("%d ", *(int*)current->data);
        current = current->next;
    }
    printf("\n");

    printf("\n=== Testing erase at index 3 ===\n");
    int* erased = erase(&list, 3);
    printf("Erased value: %d\n", *erased);
    free(erased);

    printf("\n=== Testing pop_front ===\n");
    int* front_val = pop_front(&list);
    printf("Popped front: %d\n", *front_val);
    free(front_val);

    printf("\n=== Testing pop_back ===\n");
    int* back_val = pop_back(&list);
    printf("Popped back: %d\n", *back_val);
    free(back_val);

    printf("\n=== Testing value_n_from_end (n=1) ===\n");
    int* nth = value_n_from_end(&list, 1);
    printf("1 from end: %d\n", *nth);

    printf("\n=== Testing reverse ===\n");
    reverse(&list);

    printf("List after reverse:\n");
    current = list.head;
    while (current) {
        printf("%d ", *(int*)current->data);
        current = current->next;
    }
    printf("\n");

    printf("\n=== Testing remove_value (99) ===\n");
    remove_value(&list, special);
    free(special);

    printf("List after remove_value:\n");
    current = list.head;
    while (current) {
        printf("%d ", *(int*)current->data);
        current = current->next;
    }
    printf("\n");

    printf("\n=== Cleaning up remaining nodes ===\n");
    while (!empty(&list)) {
        int* val = pop_front(&list);
        free(val);
    }

    free_list(&list);

    printf("Final size: %lu\n", (unsigned long)list.size);
    printf("All tests completed successfully.\n");

    return 0;
}

File: linked_list.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None 


class LinkedList:
    def __init__(self):
        self._size = 0
        self.head = None

    def size(self):
        return self._size 

    def empty(self):
        return self._size == 0

    def value_at(self, index):
        if index < 0 or index >= self._size:
            raise IndexError("Index out of range.")
        current = self.head 
        for i in range(index):
            current = current.next     
        return current.data  

    def push_front(self, value):
        new_node = Node(value)
        new_node.next = self.head
        self.head = new_node
        self._size += 1

    def push_back(self, value):
        new_node = Node(value)  

        if self.empty():
            self.head = new_node  
        else:
            current = self.head
            while current.next:
                current = current.next 
            current.next = new_node

        self._size += 1
                

    def pop_back(self):
        if self.empty():
           raise Exception("List is empty.") 

        if self._size == 1:
            value = self.head.data
            self.head = None
        else:
            current = self.head
            while current.next.next:
                current = current.next
            value = current.next.data
            current.next = None

        self._size -= 1
        return value


    def front(self):
        if self.empty():
            raise Exception("List is empty")
        return self.head.data


    def back(self):
        if self.empty():
            raise Exception("List is empty")

        current = self.head
        while current.next:
            current = current.next
        return current.data


    def insert(self, index, value):
        if index < 0 or index > self._size:
            raise IndexError("Index out of range")

        if index == 0:
            self.push_front(value)
            return
        
        new_node = Node(value)
        current = self.head

        for i in range(index - 1):
            current = current.next

        new_node.next = current.next
        current.next = new_node
        self._size += 1


    def erase(self, index):
        if index < 0 or index >= self._size:
            raise IndexError("Index out of range")
        
        if index == 0:
            self.head = self.head.next
        else:
            current = self.head
            for i in range(index - 1):
                current = current.next
            current.next = current.next.next

        self._size -= 1
        
    def value_n_from_end(self, n):
        if n < 0 or n >= self._size:
            raise IndexError("Index out of range")
        
        fast = self.head
        slow = self.head

        for i in range(n):
            fast = fast.next

        while fast.next:
            fast = fast.next
            slow = slow.next

        return slow.data

    def reverse(self):
        prev = None
        current = self.head

        while current:
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node

        self.head = prev

    def remove_value(self, value):
        if self.empty():
            return
        
        if self.head.data == value:
            self.head = self.head.next
            self._size -= 1
            return

        current = self.head
        while current.next:
            if current.next.data == value:
                current.next = current.next.next
                self._size -= 1
                return
            current = current.next

class LinkedListWithTail:
    def __init__(self):
        self._size = 0
        self.head = None
        self.tail = None

    def size(self):
        return self._size

    def empty(self):
        return self._size == 0

    def push_front(self, value):
        new_node = Node(value)
        new_node.next = self.head
        self.head = new_node

        if self._size == 0:
            self.tail = new_node

        self._size += 1

    def push_back(self, value):
        new_node = Node(value)

        if self.empty():
            self.head = self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node

        self._size += 1

    def pop_front(self):
        if self.empty():
            raise Exception("List is empty")

        value = self.head.data
        self.head = self.head.next
        self._size -= 1

        if self._size == 0:
            self.tail = None

        return value

    def front(self):
        if self.empty():
            raise Exception("List is empty")
        return self.head.data

    def back(self):
        if self.empty():
            raise Exception("List is empty")
        return self.tail.data



def run_tests():
    ll = LinkedList()
    assert ll.empty() == True
    assert ll.size() == 0

    ll.push_front(10)
    ll.push_front(5)
    assert ll.front() == 5
    assert ll.size() == 2

    ll.push_back(20)
    ll.push_back(30)
    assert ll.back() == 30
    assert ll.size() == 4

    assert ll.value_at(0) == 5
    assert ll.value_at(1) == 10
    assert ll.value_at(2) == 20
    assert ll.value_at(3) == 30

    ll.erase(2)
    assert ll.value_at(2) == 30
    assert ll.size() == 3

    assert ll.value_n_from_end(0) == 30
    assert ll.value_n_from_end(1) == 10
    assert ll.value_n_from_end(2) == 5

    ll.remove_value(20)
    assert ll.size() == 3
    assert ll.value_at(2) == 30

    ll.reverse()
    assert ll.value_at(0) == 30
    assert ll.value_at(1) == 10
    assert ll.value_at(2) == 5

    value = ll.pop_back()
    assert value == 5
    assert ll.size() == 2

    assert ll.front() == 30
    assert ll.back() == 10

if __name__=="__main__":
    run_tests()

